static/app/views/profiling/profileDetails/components/profileDetailsTable.tsx

```diff
@@ -0,0 +1,477 @@
+import {renderTableHead} from 'sentry/utils/profiling/tableRenderer';
+import {makeFormatter} from 'sentry/utils/profiling/units/units';
+import {useEffectAfterFirstRender} from 'sentry/utils/useEffectAfterFirstRender';
+import {useLocation} from 'sentry/utils/useLocation';
+import {useParams} from 'sentry/utils/useParams';
+
+import {useProfileGroup} from '../../profileGroupProvider';
+import {useColumnFilters} from '../hooks/useColumnFilters';
+import {useFuseSearch} from '../hooks/useFuseSearch';
+import {usePageLinks} from '../hooks/usePageLinks';
+import {useQuerystringState} from '../hooks/useQuerystringState';
+import {useSortableColumns} from '../hooks/useSortableColumn';
+import {aggregate, AggregateColumnConfig, collectProfileFrames} from '../utils';
+
+const RESULTS_PER_PAGE = 50;
+
+export function ProfileDetailsTable() {
+  const location = useLocation();
+  const [state] = useProfileGroup();
+  const [groupByViewKey, setGroupByView] = useQuerystringState({
+    key: 'detailView',
+    defaultState: 'occurrence',
+  });
+
+  const [searchQuery, setSearchQuery] = useQuerystringState({
+    key: 'query',
+    defaultState: '',
+  });
+
+  const [paginationCursor, setPaginationCursor] = useQuerystringState({
+    key: 'cursor',
+    defaultState: '',
+  });
+
+  const groupByView = GROUP_BY_OPTIONS[groupByViewKey] ?? GROUP_BY_OPTIONS.occurrence;
+
+  const cursor = parseInt(paginationCursor, 10) || 0;
+
+  const allData: TableDataRow[] = useMemo(() => {
+    const data =
+      state.type === 'resolved' ? state.data.profiles.flatMap(collectProfileFrames) : [];
+
+    return groupByView.transform(data);
+  }, [state, groupByView]);
+
+  const {search} = useFuseSearch(allData, {
+    keys: groupByView.search.key,
+    threshold: 0.3,
+  });
+
+  const debouncedSearch = useMemo(
+    () => debounce(searchString => setFilteredDataBySearch(search(searchString)), 500),
+    [search]
+  );
+
+  const [filteredDataBySearch, setFilteredDataBySearch] = useState<TableDataRow[]>(() => {
+    return search(searchQuery);
+  });
+
+  const pageLinks = usePageLinks(filteredDataBySearch, cursor);
+
+  const {filters, columnFilters, filterPredicate} = useColumnFilters(allData, [
+    'type',
+    'image',
+  ]);
+
+  const {currentSort, generateSortLink, sortCompareFn} = useSortableColumns({
+    ...groupByView.sort,
+    querystringKey: 'functionsSort',
+  });
+
+  const handleSearch = useCallback(
+    searchString => {
+      setSearchQuery(searchString);
+      setPaginationCursor(undefined);
+      debouncedSearch(searchString);
+    },
+    [setPaginationCursor, setSearchQuery, debouncedSearch]
+  );
+
+  useEffectAfterFirstRender(() => {
+    setFilteredDataBySearch(search(searchQuery));
+    // purposely omitted `searchQuery` as we only want this to run once.
+    // future search filters are called by handleSearch
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Is this more of an optimization since `useEffectAfterFirstRender` only calls this on the first render anyways?

Reply:
so `useEffectAfterFirstRender` is misleading; it's not `useEffectAfterFirstRenderOnce` its quite literally, don't run this effect on the first render, but handle it for all subsequent renders.

in our case, i want to delegate handling the search to the action of entering the search query, not as a side effect when  a querystring changes etc. it's fine to do that in the `useEffect` since thats a valid on mount effect.

for ex; i made an optimization to debounce the search input, but this hook was circumventing it. in fact we actually were calling setState twice.